docs/conditions.md

```diff
@@ -0,0 +1,73 @@
+# CR Conditions and Readiness Probe
+Conditions are..
+	   _the latest available observations of an object's state. They are
+	   an extension mechanism intended to be used when the details of an
+	   observation are not a priori known or would not apply to all
+	   instances of a given Kind._
+
+The HCO’s CR is a representation of the underlying component operators.  If the
+object exists, then all components exist.  If the object doesn’t exist, all
+components don’t exist.  However, the CR existence doesn’t help us with the
+state where the operators exist, but are they healthy?  This is where conditions
```

Comment:
I believe it's worth pointing out the `Reason` and `Message` (and maybe the whole Condition struct) to make it clear that we should be giving a reason and message for transitions into (and out of) the different ConditionTypes.

Copying the [CVO types](https://github.com/openshift/api/blob/master/config/v1/types_cluster_operator.go):

```
type ClusterOperatorStatusCondition struct {
	// type specifies the state of the operator's reconciliation functionality.
	Type ClusterStatusConditionType `json:"type"`

	// status of the condition, one of True, False, Unknown.
	Status ConditionStatus `json:"status"`

	// lastTransitionTime is the time of the last update to the current status object.
	LastTransitionTime metav1.Time `json:"lastTransitionTime"`

	// reason is the reason for the condition's last transition.  Reasons are CamelCase
	Reason string `json:"reason,omitempty"`

	// message provides additional information about the current condition.
	// This is only to be consumed by humans.
	Message string `json:"message,omitempty"`
}
```

Reply:
:+1: Let's explicitly point out that the conditions are the same condition type like in k8s (and CVO).

Reply:
Oh, and let's maybe also standardize some of the `Reson` strings: e.g. `Healing` or `Updating`. I guess we can find some in the okd repo.

Reply:
Just keep in mind that, assuming we follow the pattern set by CVO, there would be at least 3 ConditionTypes `Progressing|Available|Degraded` that can be `True|False|Unkown`. It may be difficult to standardize on the reasons for all of those but I would be interested in the proposal.

Reply:
I only talked about below `True and False` and left out `Unknown` because I wasn't sure how to write the `Meaning` section.

For instance, if you're `OperatorAvailable = true` `OperatorProgressing = unknown` and `OperatorDegraded = false`, what definitive meaning can we gather?  Does `Unknown` imply the operator doesn't know what's going on?  If that's the case then `Unknown` could be the "yellow warning light" to call for help?   Maybe that can be useful during failed upgrades...

Reply:
My comment about "Just keep in mind" was more directed at the standardizing on `Reason` string.

I don't have an immediate understanding of a scenario where the `Progressing` Condition would be `Unknown`...maybe when we first create the conditions slice? But I agree this proposal should focus on what `True|False` mean.

Reply:
> Just keep in mind that, assuming we follow the pattern set by CVO, there would be at least 3 ConditionTypes `Progressing|Available|Degraded` that can be `True|False|Unkown`. It may be difficult to standardize on the reasons for all of those but I would be interested in the proposal.

Yes you are right. Let me rephrase, maybe some suggestions would be good. For instance k8s uses `FailedCreate`, `FailedUpdate`, to indicate REST issues. But I guess it is more important to have the conditions standardized.